urllib3/connection.py

```diff
@@ -299,7 +302,15 @@ def connect(self):
                     'for details.)'.format(hostname)),
                     SubjectAltNameWarning
                 )
-            match_hostname(cert, self.assert_hostname or hostname)
+            try:
+                match_hostname(cert, self.assert_hostname or hostname)
+            except CertificateError as e:
+                log.error(
+                    'Certificate did not match expected hostname: %s. '
+                    'Certificate: %s', asserted_hostname, cert
+                )
+                e.cert = cert
+                raise e
```

Comment:
That's definitely another alternative. I'm a bit sad about that though: it'd be convenient to have it.

You could investigate whether leaving the patch in and changing to a naked `raise` statement still successfully patches the cert onto the exception. That would be the best of both worlds I think.